<a href="https://colab.research.google.com/github/Luiz-Malanczyn/BCC_TCC_PSI/blob/main/relatorios%20/TCC_desempenho_algortimos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Proposta de Pesquisa Científica:
##Um Método de Detecção de Mudança baseado em PSI
###Gabriel Luiz Tokarski Soto, Luiz Eduardo Malanczyn de Oliveira
###Orientador: Fabrício Enembreck  
\
##TCC - SPRINT 1
##Relatório dos testes

###Preparação dos dados

####Importando as bibliotecas utilizadas

In [72]:
import numpy as np
import pandas as pd

####Preparando os arquivos csv disponiveis no repositório da equipe

In [73]:
#Caminho do csv contendo os resultados da execução dos testes realizados no MOA através do algoritmo automatizador.
#O primeiro arquivo representa os dados dos testes de tuning para os parametros dos thresholds.
!wget https://raw.githubusercontent.com/Luiz-Malanczyn/BCC_TCC_PSI/main/csv_tests/tuningThresholds_clean.csv -O thresholds.csv
!wget https://raw.githubusercontent.com/Luiz-Malanczyn/BCC_TCC_PSI/main/csv_tests/tuningWindow_clean.csv -O results_window.csv
!wget https://raw.githubusercontent.com/Luiz-Malanczyn/BCC_TCC_PSI/main/csv_tests/otherDetectors_clean.csv -O results_others_clean.csv
!wget https://raw.githubusercontent.com/Luiz-Malanczyn/BCC_TCC_PSI/main/csv_tests/finalTuning_clean.csv -O result_tuned_clean.csv

path = '/content/thresholds.csv'
path_window = '/content/results_window.csv'
path_others = '/content/results_others_clean.csv'
path_tuned = '/content/result_tuned_clean.csv'

--2023-09-19 03:56:27--  https://raw.githubusercontent.com/Luiz-Malanczyn/BCC_TCC_PSI/main/csv_tests/tuningThresholds_clean.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3384 (3.3K) [text/plain]
Saving to: ‘thresholds.csv’

thresholds.csv      100%[===================>]   3.30K  --.-KB/s    in 0s      

2023-09-19 03:56:27 (43.8 MB/s) - ‘thresholds.csv’ saved [3384/3384]

--2023-09-19 03:56:27--  https://raw.githubusercontent.com/Luiz-Malanczyn/BCC_TCC_PSI/main/csv_tests/tuningWindow_clean.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response..

&emsp;Os testes foram realizados com 2 conjuntos de streams geradas através do MOA, ambos de mudança abrupta e gradual, o primeiro par de streams, utilizado no tuning do PSI foi criado utilizando uma seed rândomica, enquanto o segundo, utilizado na comparação dos detectores, foi criado com outra.

###Primeiro arquivo, testes de tuning para os parametros de threshold de detecção de aviso e de mudança

In [74]:
df = pd.read_csv(path)

#Tratativa para incluir índices representativos do que cada linha representa, sendo eles os parâmetros utilizados nos testes
contador = 0
parametros = ['(2,5)',
'(2,10)',
'(2, 15)',
'(2, 20)',
'(2, 30)',
'(5, 10)',
'(5, 15)',
'(5, 20)',
'(5,30)',
'(10, 15)',
'(10, 20)',
'(10,30)',
'(15, 20)',
'(15, 30)',
'(20, 30)']

novo_indice = []
tipo = ''
for i in df.index:
    if i % 2 == 0:
      tipo = 'Abrupt'
      if i != 0:
        contador+=1
    else:
      tipo = 'Gradual'
    param = tipo + parametros[contador]
    novo_indice.append(param)

df.index = novo_indice
#Convertendo os dados do dataframe para poder realizar eventuais cálculos e comparações dos valores
df = df.astype(float)
df

Learning evaluation instances  Evaluation time    Model cost  \
Abrupt(2,5)                           100000.0         0.203125  4.485547e-10   
Gradual(2,5)                          100000.0         0.250000  5.474107e-10   
Abrupt(2,10)                          100000.0         0.437500  9.634015e-10   
Gradual(2,10)                         100000.0         0.359375  1.170685e-09   
Abrupt(2, 15)                         100000.0         0.421875  1.389999e-09   
Gradual(2, 15)                        100000.0         0.390625  1.287036e-09   
Abrupt(2, 20)                         100000.0         0.437500  1.441480e-09   
Gradual(2, 20)                        100000.0         0.437500  1.441480e-09   
Abrupt(2, 30)                         100000.0         0.437500  1.441480e-09   
Gradual(2, 30)                        100000.0         0.421875  1.389999e-09   
Abrupt(5, 10)                         100000.0         0.343750  7.569583e-10   
Gradual(5, 10)                        100000.0         0.359375  1.170685e-09   
Abrupt(5, 15)                         100000.0         0.437500  1.441480e-09   
Gradual(5, 15)                        100000.0         0.359375  1.184073e-09   
Abrupt(5, 20)                         100000.0         0.406250  1.338518e-09   
Gradual(5, 20)                        100000.0         0.406250  1.338518e-09   
Abrupt(5,30)                          100000.0         0.406250  1.338518e-09   
Gradual(5,30)                         100000.0         0.421875  1.389999e-09   
Abrupt(10, 15)                        100000.0         0.484375  1.595925e-09   
Gradual(10, 15)                       100000.0         0.421875  1.389999e-09   
Abrupt(10, 20)                        100000.0         0.453125  1.492962e-09   
Gradual(10, 20)                       100000.0         0.421875  1.389999e-09   
Abrupt(10,30)                         100000.0         0.406250  1.338518e-09   
Gradual(10,30)                        100000.0         0.406250  1.338518e-09   
Abrupt(15, 20)                        100000.0         0.437500  1.441480e-09   
Gradual(15, 20)                       100000.0         0.437500  1.441480e-09   
Abrupt(15, 30)                        100000.0         0.406250  1.338518e-09   
Gradual(15, 30)                       100000.0         0.421875  1.389999e-09   
Abrupt(20, 30)                        100000.0         0.406250  1.338518e-09   
Gradual(20, 30)                       100000.0         0.437500  1.441480e-09   

                 Learned instances  Detected changes  Detected Warnings  \
Abrupt(2,5)               100000.0             476.0                0.0   
Gradual(2,5)              100000.0             480.0                0.0   
Abrupt(2,10)              100000.0             343.0                1.0   
Gradual(2,10)             100000.0             349.0                0.0   
Abrupt(2, 15)             100000.0             277.0                0.0   
Gradual(2, 15)            100000.0             281.0                0.0   
Abrupt(2, 20)             100000.0             260.0                0.0   
Gradual(2, 20)            100000.0             264.0                0.0   
Abrupt(2, 30)             100000.0             259.0                0.0   
Gradual(2, 30)            100000.0             261.0                0.0   
Abrupt(5, 10)             100000.0             343.0                1.0   
Gradual(5, 10)            100000.0             349.0                0.0   
Abrupt(5, 15)             100000.0             277.0                0.0   
Gradual(5, 15)            100000.0             281.0                0.0   
Abrupt(5, 20)             100000.0             260.0                0.0   
Gradual(5, 20)            100000.0             264.0                0.0   
Abrupt(5,30)              100000.0             259.0                0.0   
Gradual(5,30)             100000.0             261.0                0.0   
Abrupt(10, 15)            100000.0             277.0                0.0   
Gradu

&emsp;Separando os dados úteis para a avaliação em um novo dataframe, os dados escolhidos são:
- A quantidade de mudanças detectadas pelo algoritmo
- A quantidade de avisos detectados pelo algoritmo
- A quantidade de mudanças verdadeiras na stream
- O tempo de atraso para detecção
\
&emsp;Também foi adicionado um campo que representa qual a porcentagem das mudanças detectadas eram de fato mudanças verdadeiras.

In [98]:
#Separando as colunas relevantes no novo dataframe
new_df = df[['Detected changes', 'Detected Warnings', 'True changes', 'Delay detection', 'True changes detected']]

#Adicionando novo campo representando quantos dos acertos detectados eram mudanças de conceito reais.
accuracy = (new_df['True changes']/new_df['Detected changes']) * 100
new_df.loc[:,'True Change Accuracy'] = accuracy
new_df

<ipython-input-98-d595b28f9929>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df.loc[:,'True Change Accuracy'] = accuracy


Detected changes  Detected Warnings  True changes  \
Abrupt(2,5)                 476.0                0.0          10.0   
Gradual(2,5)                480.0                0.0          10.0   
Abrupt(2,10)                343.0                1.0          10.0   
Gradual(2,10)               349.0                0.0          10.0   
Abrupt(2, 15)               277.0                0.0          10.0   
Gradual(2, 15)              281.0                0.0          10.0   
Abrupt(2, 20)               260.0                0.0          10.0   
Gradual(2, 20)              264.0                0.0          10.0   
Abrupt(2, 30)               259.0                0.0          10.0   
Gradual(2, 30)              261.0                0.0          10.0   
Abrupt(5, 10)               343.0                1.0          10.0   
Gradual(5, 10)              349.0                0.0          10.0   
Abrupt(5, 15)               277.0                0.0          10.0   
Gradual(5, 15)              281.0                0.0          10.0   
Abrupt(5, 20)               260.0                0.0          10.0   
Gradual(5, 20)              264.0                0.0          10.0   
Abrupt(5,30)                259.0                0.0          10.0   
Gradual(5,30)               261.0                0.0          10.0   
Abrupt(10, 15)              277.0                0.0          10.0   
Gradual(10, 15)             281.0                0.0          10.0   
Abrupt(10, 20)              260.0                0.0          10.0   
Gradual(10, 20)             264.0                0.0          10.0   
Abrupt(10,30)               259.0                0.0          10.0   
Gradual(10,30)              261.0                0.0          10.0   
Abrupt(15, 20)              260.0                0.0          10.0   
Gradual(15, 20)             264.0                0.0          10.0   
Abrupt(15, 30)              259.0                0.0          10.0   
Gradual(15, 30)             261.0                0.0          10.0   
Abrupt(20, 30)              259.0                0.0          10.0   
Gradual(20, 30)             261.0                0.0          10.0   

                 Delay detection  True changes detected  True Change Accuracy  
Abrupt(2,5)                 59.1                   10.0              2.100840  
Gradual(2,5)                46.0                   10.0              2.083333  
Abrupt(2,10)                32.5                   10.0              2.915452  
Gradual(2,10)               33.5                   10.0              2.865330  
Abrupt(2, 15)               10.2                   10.0              3.610108  
Gradual(2, 15)              18.3                   10.0              3.558719  
Abrupt(2, 20)               12.4                   10.0              3.846154  
Gradual(2, 20)              21.4                   10.0              3.787879  
Abrupt(2, 30)               20.1                   10.0              3.861004  
Gradual(2, 30)              28.1                   10.0              3.831418  
Abrupt(5, 10)               32.5                   10.0              2.915452  
Gradual(5, 10)              33.5                   10.0              2.865330  
Abrupt(5, 15)               10.2                   10.0              3.610108  
Gradual(5, 15)              18.3                   10.0              3.558719  
Abrupt(5, 20)               12.4                   10.0              3.846154  
Gradual(5, 20)              21.4                   10.0              3.787879  
Abrupt(5,30)                20.1                   10.0              3.861004  
Gradual(5,30)               28.1                   10.0              3.831418  
Abrupt(10, 15)              10.2                   10.0              3.610108  
Gradual(10, 15)             18.3                   10.0              3.558719  
Abrupt(10, 20)              12.4                   10.0              3.846154  
Gradual(10, 20)             21.4                   10.0              3.787879  
Abrupt

&emsp;Com base nos resultados acima, foi escolhido aplicar um filtro, mantendo assim somente os casos onde a acurácia foi superior a média da coluna 'True Change Accuracy',

In [76]:
media = np.mean(new_df['True Change Accuracy'])
media

3.541681864749186

In [77]:
min_acc = (new_df['True Change Accuracy'] >= media)
min_acc_df = new_df[min_acc]
min_acc_df = min_acc_df.sort_values(by='True Change Accuracy', ascending=False)
min_acc_df

Detected changes  Detected Warnings  True changes  \
Abrupt(10,30)               259.0                0.0          10.0   
Abrupt(2, 30)               259.0                0.0          10.0   
Abrupt(20, 30)              259.0                0.0          10.0   
Abrupt(15, 30)              259.0                0.0          10.0   
Abrupt(5,30)                259.0                0.0          10.0   
Abrupt(2, 20)               260.0                0.0          10.0   
Abrupt(10, 20)              260.0                0.0          10.0   
Abrupt(5, 20)               260.0                0.0          10.0   
Abrupt(15, 20)              260.0                0.0          10.0   
Gradual(20, 30)             261.0                0.0          10.0   
Gradual(5,30)               261.0                0.0          10.0   
Gradual(10,30)              261.0                0.0          10.0   
Gradual(15, 30)             261.0                0.0          10.0   
Gradual(2, 30)              261.0                0.0          10.0   
Gradual(15, 20)             264.0                0.0          10.0   
Gradual(10, 20)             264.0                0.0          10.0   
Gradual(2, 20)              264.0                0.0          10.0   
Gradual(5, 20)              264.0                0.0          10.0   
Abrupt(2, 15)               277.0                0.0          10.0   
Abrupt(5, 15)               277.0                0.0          10.0   
Abrupt(10, 15)              277.0                0.0          10.0   
Gradual(10, 15)             281.0                0.0          10.0   
Gradual(2, 15)              281.0                0.0          10.0   
Gradual(5, 15)              281.0                0.0          10.0   

                 Delay detection  True changes detected  True Change Accuracy  
Abrupt(10,30)               20.1                   10.0              3.861004  
Abrupt(2, 30)               20.1                   10.0              3.861004  
Abrupt(20, 30)              20.1                   10.0              3.861004  
Abrupt(15, 30)              20.1                   10.0              3.861004  
Abrupt(5,30)                20.1                   10.0              3.861004  
Abrupt(2, 20)               12.4                   10.0              3.846154  
Abrupt(10, 20)              12.4                   10.0              3.846154  
Abrupt(5, 20)               12.4                   10.0              3.846154  
Abrupt(15, 20)              12.4                   10.0              3.846154  
Gradual(20, 30)             28.1                   10.0              3.831418  
Gradual(5,30)               28.1                   10.0              3.831418  
Gradual(10,30)              28.1                   10.0              3.831418  
Gradual(15, 30)             28.1                   10.0              3.831418  
Gradual(2, 30)              28.1                   10.0              3.831418  
Gradual(15, 20)             21.4                   10.0              3.787879  
Gradual(10, 20)             21.4                   10.0              3.787879  
Gradual(2, 20)              21.4                   10.0              3.787879  
Gradual(5, 20)              21.4                   10.0              3.787879  
Abrupt(2, 15)               10.2                   10.0              3.610108  
Abrupt(5, 15)               10.2                   10.0              3.610108  
Abrupt(10, 15)              10.2                   10.0              3.610108  
Gradual(10, 15)             18.3                   10.0              3.558719  
Gradual(2, 15)              18.3                   10.0              3.558719  
Gradual(5, 15)              18.3                   10.0              3.558719

Ordenando os resultados pela taxa de precisão de mudanças verdadeiras, é possível perceber que as primeiras posições estão ocupadas pelos testes com parametro de valor de detecção de mudança igual a 30, visualizando através dos 5 primeiros resultados conforme a tabela seguinte

In [78]:
min_acc_df.head(5)

Detected changes  Detected Warnings  True changes  \
Abrupt(10,30)              259.0                0.0          10.0   
Abrupt(2, 30)              259.0                0.0          10.0   
Abrupt(20, 30)             259.0                0.0          10.0   
Abrupt(15, 30)             259.0                0.0          10.0   
Abrupt(5,30)               259.0                0.0          10.0   

                Delay detection  True changes detected  True Change Accuracy  
Abrupt(10,30)              20.1                   10.0              3.861004  
Abrupt(2, 30)              20.1                   10.0              3.861004  
Abrupt(20, 30)             20.1                   10.0              3.861004  
Abrupt(15, 30)             20.1                   10.0              3.861004  
Abrupt(5,30)               20.1                   10.0              3.861004

In [79]:
acc = min_acc_df.loc['Gradual(2, 30)', 'True Change Accuracy']
print(f"True Change Accuracy para Gradual(2, 30): {acc}")

True Change Accuracy para Gradual(2, 30): 3.8314176245210727


Os dois 'piores' resultados para a taxa de acurácia foram obtidos pelos testes com parametros (2, 30) e (15, 30) ambos com streams graduais. Mesmo assim se mantiveram acima da taxa de precisão de 3.83%, valor acima da média das amostras.

Sendo assim o conjunto de pares contendo o valor 30 para a mudança detectada foi escolhido para refinar a parametrização do tamanho da janela, visto que com tamanho 100 não foi possível detectar situções de aviso, e nem identificar corretamente a ocorrência de mudanças de conceito.

###Segundo arquivo, testes de tuning para o arquivo com os testes de tamanho da janela

&emsp;Com os resultados obtidos no teste anterior, foram definidos parâmetros fixos correspondentes aos melhores resultados para determinar o melhor valor possivel para o tamanho da janela.

In [80]:
df_window = pd.read_csv(path_window)
contador = 0

#Assim como no teste anterior, aplica-se uma tratativa para nomear os índices conforme os parametros testados
parametros = [
'(100, 2, 30)',
'(100, 5,30)',
'(100, 10,30)',
'(100, 15, 30)',
'(100, 20, 30)',
'(500, 2, 30)',
'(500, 5,30)',
'(500, 10,30)',
'(500, 15, 30)',
'(500, 20, 30)',
'(1000, 2, 30)',
'(1000, 5,30)',
'(1000, 10,30)',
'(1000, 15, 30)',
'(1000, 20, 30)'
]

novo_indice = []
tipo = ''
for i in df_window.index:
    if i % 2 == 0:
      tipo = 'Abrupt'
      if i != 0:
        contador+=1
    else:
      tipo = 'Gradual'
    param = tipo + parametros[contador]
    novo_indice.append(param)

df_window.index = novo_indice
df_window = df_window.astype(float)
df_window

Learning evaluation instances  Evaluation time  \
Abrupt(100, 2, 30)                          100000.0         0.328125   
Gradual(100, 2, 30)                         100000.0         0.265625   
Abrupt(100, 5,30)                           100000.0         0.281250   
Gradual(100, 5,30)                          100000.0         0.359375   
Abrupt(100, 10,30)                          100000.0         0.203125   
Gradual(100, 10,30)                         100000.0         0.218750   
Abrupt(100, 15, 30)                         100000.0         0.343750   
Gradual(100, 15, 30)                        100000.0         0.250000   
Abrupt(100, 20, 30)                         100000.0         0.250000   
Gradual(100, 20, 30)                        100000.0         0.265625   
Abrupt(500, 2, 30)                          100000.0         1.281250   
Gradual(500, 2, 30)                         100000.0         1.140625   
Abrupt(500, 5,30)                           100000.0         1.250000   
Gradual(500, 5,30)                          100000.0         1.125000   
Abrupt(500, 10,30)                          100000.0         1.281250   
Gradual(500, 10,30)                         100000.0         1.140625   
Abrupt(500, 15, 30)                         100000.0         1.250000   
Gradual(500, 15, 30)                        100000.0         1.078125   
Abrupt(500, 20, 30)                         100000.0         1.359375   
Gradual(500, 20, 30)                        100000.0         1.031250   
Abrupt(1000, 2, 30)                         100000.0         2.109375   
Gradual(1000, 2, 30)                        100000.0         2.156250   
Abrupt(1000, 5,30)                          100000.0         2.234375   
Gradual(1000, 5,30)                         100000.0         2.015625   
Abrupt(1000, 10,30)                         100000.0         1.921875   
Gradual(1000, 10,30)                        100000.0         2.140625   
Abrupt(1000, 15, 30)                        100000.0         2.187500   
Gradual(1000, 15, 30)                       100000.0         1.906250   
Abrupt(1000, 20, 30)                        100000.0         2.109375   
Gradual(1000, 20, 30)                       100000.0         2.218750   

                         Model cost  Learned instances  Detected changes  \
Abrupt(100, 2, 30)     1.081110e-09           100000.0             259.0   
Gradual(100, 2, 30)    8.751845e-10           100000.0             261.0   
Abrupt(100, 5,30)      9.266660e-10           100000.0             259.0   
Gradual(100, 5,30)     1.184073e-09           100000.0             261.0   
Abrupt(100, 10,30)     6.692588e-10           100000.0             259.0   
Gradual(100, 10,30)    7.207402e-10           100000.0             261.0   
Abrupt(100, 15, 30)    1.132592e-09           100000.0             259.0   
Gradual(100, 15, 30)   8.237031e-10           100000.0             261.0   
Abrupt(100, 20, 30)    8.237031e-10           100000.0             259.0   
Gradual(100, 20, 30)   8.751845e-10           100000.0             261.0   
Abrupt(500, 2, 30)     1.187158e-08           100000.0              35.0   
Gradual(500, 2, 30)    1.056860e-08           100000.0              44.0   
Abrupt(500, 5,30)      1.158203e-08           100000.0              35.0   
Gradual(500, 5,30)     1.042383e-08           100000.0              44.0   
Abrupt(500, 10,30)     1.187158e-08           100000.0              35.0   
Gradual(500, 10,30)    1.056860e-08           100000.0              44.0   
Abrupt(500, 15, 30)    1.158203e-08           100000.0              35.0   
Gradual(500, 15, 30)   9.989502e-09           100000.0              44.0   
Abrupt(500, 20, 30)    1.259546e-08           100000.0              35.0   
Gradual(500, 20, 30)   9.555176e-09           100000.0              44.0   
Abrupt(1000, 2, 30)    3.450550e-08           100000.0              19.0   
Gradual(1000, 2, 30)   3.527229e-08           100000.0              19.0   
Abrupt(1000

&emsp;Separando do novo dataframe somente os dados relevantes como mencionado anteriormente na tratativa do primeiro teste, obtem-se os dados do teste do tamanho da janela

In [81]:
new_df_window = df_window[['Detected changes', 'Detected Warnings', 'True changes', 'Delay detection', 'True changes detected']]
accuracy = (new_df_window['True changes']/new_df_window['Detected changes']) * 100
new_df_window.loc[:,'True Change Accuracy'] = accuracy
new_df_window

<ipython-input-81-441461128114>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df_window.loc[:,'True Change Accuracy'] = accuracy


Detected changes  Detected Warnings  True changes  \
Abrupt(100, 2, 30)                259.0                0.0          10.0   
Gradual(100, 2, 30)               261.0                0.0          10.0   
Abrupt(100, 5,30)                 259.0                0.0          10.0   
Gradual(100, 5,30)                261.0                0.0          10.0   
Abrupt(100, 10,30)                259.0                0.0          10.0   
Gradual(100, 10,30)               261.0                0.0          10.0   
Abrupt(100, 15, 30)               259.0                0.0          10.0   
Gradual(100, 15, 30)              261.0                0.0          10.0   
Abrupt(100, 20, 30)               259.0                0.0          10.0   
Gradual(100, 20, 30)              261.0                0.0          10.0   
Abrupt(500, 2, 30)                 35.0                7.0          10.0   
Gradual(500, 2, 30)                44.0               10.0          10.0   
Abrupt(500, 5,30)                  35.0               15.0          10.0   
Gradual(500, 5,30)                 44.0                9.0          10.0   
Abrupt(500, 10,30)                 35.0               14.0          10.0   
Gradual(500, 10,30)                44.0               21.0          10.0   
Abrupt(500, 15, 30)                35.0               19.0          10.0   
Gradual(500, 15, 30)               44.0               17.0          10.0   
Abrupt(500, 20, 30)                35.0               13.0          10.0   
Gradual(500, 20, 30)               44.0               17.0          10.0   
Abrupt(1000, 2, 30)                19.0               10.0          10.0   
Gradual(1000, 2, 30)               19.0               10.0          10.0   
Abrupt(1000, 5,30)                 19.0               18.0          10.0   
Gradual(1000, 5,30)                19.0               19.0          10.0   
Abrupt(1000, 10,30)                19.0               22.0          10.0   
Gradual(1000, 10,30)               19.0               20.0          10.0   
Abrupt(1000, 15, 30)               19.0               15.0          10.0   
Gradual(1000, 15, 30)              19.0               21.0          10.0   
Abrupt(1000, 20, 30)               19.0               16.0          10.0   
Gradual(1000, 20, 30)              19.0               11.0          10.0   

                       Delay detection  True changes detected  \
Abrupt(100, 2, 30)                20.1                   10.0   
Gradual(100, 2, 30)               28.1                   10.0   
Abrupt(100, 5,30)                 20.1                   10.0   
Gradual(100, 5,30)                28.1                   10.0   
Abrupt(100, 10,30)                20.1                   10.0   
Gradual(100, 10,30)               28.1                   10.0   
Abrupt(100, 15, 30)               20.1                   10.0   
Gradual(100, 15, 30)              28.1                   10.0   
Abrupt(100, 20, 30)               20.1                   10.0   
Gradual(100, 20, 30)              28.1                   10.0   
Abrupt(500, 2, 30)               101.1                   10.0   
Gradual(500, 2, 30)              806.5                   10.0   
Abrupt(500, 5,30)                101.1                   10.0   
Gradual(500, 5,30)               806.5                   10.0   
Abrupt(500, 10,30)               101.1                   10.0   
Gradual(500, 10,30)              806.5                   10.0   
Abrupt(500, 15, 30)              101.1                   10.0   
Gradual(500, 15, 30)             806.5                   10.0   
Abrupt(500, 20, 30)              101.1                   10.0   
Gradual(500, 20, 30)             806.5                   10.0   
Abrupt(1000, 2, 30)               85.7                   10.0   
Gradual(1000, 2, 30)              96.3                   10.0   
Abrupt(1000, 5,30)                85.7                   10.0   
Gradual(1000, 5,30)               96.3                   10.0   
Abrupt(1000, 10,30)               85.7   

&emsp;Para os casos de janelas com tamanho 100, novamente os resultados não foram bons e além de não gerarem avisos, detectaram um valor muito acima do real em relação a quantidade de mudanças reais na stream.

&emsp;Porém para os casos de janela igual a 500, apesar da melhora em relação a detecção de falsos positivos, obtendo uma quantidade mais próxima da real, o atraso para a detecção aumentou significativamente.

In [82]:
tunning_final = (new_df_window['True Change Accuracy'] >= 50)
tunning_final_df = new_df_window[tunning_final]
tunning_final_df.head(10)

Detected changes  Detected Warnings  True changes  \
Abrupt(1000, 2, 30)                19.0               10.0          10.0   
Gradual(1000, 2, 30)               19.0               10.0          10.0   
Abrupt(1000, 5,30)                 19.0               18.0          10.0   
Gradual(1000, 5,30)                19.0               19.0          10.0   
Abrupt(1000, 10,30)                19.0               22.0          10.0   
Gradual(1000, 10,30)               19.0               20.0          10.0   
Abrupt(1000, 15, 30)               19.0               15.0          10.0   
Gradual(1000, 15, 30)              19.0               21.0          10.0   
Abrupt(1000, 20, 30)               19.0               16.0          10.0   
Gradual(1000, 20, 30)              19.0               11.0          10.0   

                       Delay detection  True changes detected  \
Abrupt(1000, 2, 30)               85.7                   10.0   
Gradual(1000, 2, 30)              96.3                   10.0   
Abrupt(1000, 5,30)                85.7                   10.0   
Gradual(1000, 5,30)               96.3                   10.0   
Abrupt(1000, 10,30)               85.7                   10.0   
Gradual(1000, 10,30)              96.3                   10.0   
Abrupt(1000, 15, 30)              85.7                   10.0   
Gradual(1000, 15, 30)             96.3                   10.0   
Abrupt(1000, 20, 30)              85.7                   10.0   
Gradual(1000, 20, 30)             96.3                   10.0   

                       True Change Accuracy  
Abrupt(1000, 2, 30)               52.631579  
Gradual(1000, 2, 30)              52.631579  
Abrupt(1000, 5,30)                52.631579  
Gradual(1000, 5,30)               52.631579  
Abrupt(1000, 10,30)               52.631579  
Gradual(1000, 10,30)              52.631579  
Abrupt(1000, 15, 30)              52.631579  
Gradual(1000, 15, 30)             52.631579  
Abrupt(1000, 20, 30)              52.631579  
Gradual(1000, 20, 30)             52.631579

&emsp;Observa-se que nos testes onde o tamanho da janela é igual a 1000, obtêm-se resultados melhores. Os testes apresentaram detecção de mudança de conceito constante que não estava presente nos testes com janelas diferentes, apresentando um resultado condizente com o valor utilizado como parametro para detecção de mudanças.

&emsp;Além disso, os tempos de atraso para detecção não mudaram entre os diferentes valores de threshold para aviso e detecção real de mudança, mas sim entre os tipos de stream, apresentando tempos constantes de 96.3 para a stream gradual e um tempo de 85.7	para a abrupta.

&emsp;Foi escolhida como parametrização padrão final, o tamanho da janela sendo igual a 1000, o limite de aviso igual a 20 e o alerta de  detecção igual a 30. Com o valor do aviso podendo alterar conforme contexto onde for utilizado e sem prejudicar o resultado.

###Testes comparativos com os outros detectores

&emsp;Conforme comentado no teste anterior, foi escolhido o melhor parametro para realizar testes com o PSI para comparar com os principais detectores de mudança de conceito, sendo esses DDM, EDDM, ADWIN, CUSUM e Page-Hinkley.

In [85]:
#Aqui fazemos a tratativa dos índices para facilitar a leitura e identificação dos dados nos dataframes

#Principais algoritmos com parametros default
df_others = pd.read_csv(path_others)

contador = 0
parametros = [
'(DDM)',
'(EDDM)',
'(ADWIN)',
'(CUSUM)',
'(PAGE-HINKLEY)'
]

novo_indice = []
tipo = ''
for i in df_others.index:
    if i % 2 == 0:
      tipo = 'Abrupt'
      if i != 0:
        contador+=1
    else:
      tipo = 'Gradual'
    param = tipo + parametros[contador]
    novo_indice.append(param)

df_others.index = novo_indice
df_others = df_others.astype(float)

#PSI com os parametros selecionados
df_tuned = pd.read_csv(path_tuned)

contador = 0
parametros = [
'(PSI 1000, 20, 30)',
]

novo_indice = []
tipo = ''
for i in df_tuned.index:
    if i % 2 == 0:
      tipo = 'Abrupt'
      if i != 0:
        contador+=1
    else:
      tipo = 'Gradual'
    param = tipo + parametros[contador]
    novo_indice.append(param)

df_tuned.index = novo_indice
df_tuned = df_tuned.astype(float)

# junção dos dataframes
df_detectors = pd.concat([df_others, df_tuned])
df_detectors

Learning evaluation instances  Evaluation time  \
Abrupt(DDM)                                     100000.0         0.312500   
Gradual(DDM)                                    100000.0         0.343750   
Abrupt(EDDM)                                    100000.0         0.296875   
Gradual(EDDM)                                   100000.0         0.265625   
Abrupt(ADWIN)                                   100000.0         0.359375   
Gradual(ADWIN)                                  100000.0         0.406250   
Abrupt(CUSUM)                                   100000.0         0.296875   
Gradual(CUSUM)                                  100000.0         0.281250   
Abrupt(PAGE-HINKLEY)                            100000.0         0.328125   
Gradual(PAGE-HINKLEY)                           100000.0         0.406250   
Abrupt(PSI 1000, 20, 30)                        100000.0         2.531250   
Gradual(PSI 1000, 20, 30)                       100000.0         4.578125   

                             Model cost  Learned instances  Detected changes  \
Abrupt(DDM)                5.096404e-10           100000.0              10.0   
Gradual(DDM)               5.606044e-10           100000.0              12.0   
Abrupt(EDDM)               4.048990e-10           100000.0              68.0   
Gradual(EDDM)              3.622780e-10           100000.0              81.0   
Abrupt(ADWIN)              6.812883e-10           100000.0              29.0   
Gradual(ADWIN)             7.701520e-10           100000.0              26.0   
Abrupt(CUSUM)              4.860016e-10           100000.0              11.0   
Gradual(CUSUM)             4.604226e-10           100000.0              11.0   
Abrupt(PAGE-HINKLEY)       5.670396e-10           100000.0               9.0   
Gradual(PAGE-HINKLEY)      7.020491e-10           100000.0               8.0   
Abrupt(PSI 1000, 20, 30)   4.140660e-08           100000.0              20.0   
Gradual(PSI 1000, 20, 30)  7.488972e-08           100000.0              19.0   

                           Detected Warnings  Predicion error  True changes  \
Abrupt(DDM)                             11.0         0.435887          10.0   
Gradual(DDM)                            13.0         0.433484          10.0   
Abrupt(EDDM)                            10.0         1.323602          10.0   
Gradual(EDDM)                           10.0         1.321452          10.0   
Abrupt(ADWIN)                            0.0         0.328938          10.0   
Gradual(ADWIN)                           0.0         0.327279          10.0   
Abrupt(CUSUM)                            0.0         0.434162          10.0   
Gradual(CUSUM)                           0.0         0.429756          10.0   
Abrupt(PAGE-HINKLEY)                     0.0         0.442897          10.0   
Gradual(PAGE-HINKLEY)                    0.0         0.453838          10.0   
Abrupt(PSI 1000, 20, 30)                10.0         0.497350          10.0   
Gradual(PSI 1000, 20, 30)               22.0         0.500470          10.0   

                           Delay detection  True changes detected  \
Abrupt(DDM)                          305.6                   10.0   
Gradual(DDM)                         325.8                   10.0   
Abrupt(EDDM)                        1164.7                   10.0   
Gradual(EDDM)                       1124.4                   10.0   
Abrupt(ADWIN)                         42.6                   10.0   
Gradual(ADWIN)                        51.0                   10.0   
Abrupt(CUSUM)                        152.9                   10.0   
Gradual(CUSUM)                       153.3                   10.0   
Abrupt(PAGE-HINKLEY)                2253.3                    9.0   
Gradual(PAGE-HINKLEY)               2121.3                    8.0   
Abrupt(PSI 1000, 20, 30)             198.4                   10.0   
Gradual(PSI 1000, 20, 30)            101.3                   10.0   

                           Input values  Model training in

&emsp;A seguir, obtém-se o novo dataframe de comparação dos detectores, filtrando os campos desejados conforme feito nos testes anteriores.

In [107]:
new_df_detectors = df_detectors[['Detected changes', 'Detected Warnings', 'True changes', 'Delay detection', 'True changes detected']]
# accuracy = (new_df_detectors['True changes']/new_df_detectors['Detected changes']) * 100


new_df_detectors.loc[:,'True Change Accuracy'] = accuracy
new_df_detectors

<ipython-input-107-5223d384bb45>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df_detectors.loc[:,'True Change Accuracy'] = accuracy


Detected changes  Detected Warnings  True changes  \
Abrupt(DDM)                            10.0               11.0          10.0   
Gradual(DDM)                           12.0               13.0          10.0   
Abrupt(EDDM)                           68.0               10.0          10.0   
Gradual(EDDM)                          81.0               10.0          10.0   
Abrupt(ADWIN)                          29.0                0.0          10.0   
Gradual(ADWIN)                         26.0                0.0          10.0   
Abrupt(CUSUM)                          11.0                0.0          10.0   
Gradual(CUSUM)                         11.0                0.0          10.0   
Abrupt(PAGE-HINKLEY)                    9.0                0.0          10.0   
Gradual(PAGE-HINKLEY)                   8.0                0.0          10.0   
Abrupt(PSI 1000, 20, 30)               20.0               10.0          10.0   
Gradual(PSI 1000, 20, 30)              19.0               22.0          10.0   

                           Delay detection  True changes detected  \
Abrupt(DDM)                          305.6                   10.0   
Gradual(DDM)                         325.8                   10.0   
Abrupt(EDDM)                        1164.7                   10.0   
Gradual(EDDM)                       1124.4                   10.0   
Abrupt(ADWIN)                         42.6                   10.0   
Gradual(ADWIN)                        51.0                   10.0   
Abrupt(CUSUM)                        152.9                   10.0   
Gradual(CUSUM)                       153.3                   10.0   
Abrupt(PAGE-HINKLEY)                2253.3                    9.0   
Gradual(PAGE-HINKLEY)               2121.3                    8.0   
Abrupt(PSI 1000, 20, 30)             198.4                   10.0   
Gradual(PSI 1000, 20, 30)            101.3                   10.0   

                           True Change Accuracy  
Abrupt(DDM)                          100.000000  
Gradual(DDM)                          83.333333  
Abrupt(EDDM)                          14.705882  
Gradual(EDDM)                         12.345679  
Abrupt(ADWIN)                         34.482759  
Gradual(ADWIN)                        38.461538  
Abrupt(CUSUM)                         90.909091  
Gradual(CUSUM)                        90.909091  
Abrupt(PAGE-HINKLEY)                 111.111111  
Gradual(PAGE-HINKLEY)                125.000000  
Abrupt(PSI 1000, 20, 30)              50.000000  
Gradual(PSI 1000, 20, 30)             52.631579

Referente à coluna 'True Change Accuracy', como a mesma representa a quantidade, em porcentagem, de quantas mudanças detectadas de fato são mudanças verdadeiras na stream, os dois casos do PAGE-HINKLEY apresentaram valores superiores a 100% pois, na verdade, deixaram de detectar mudanças de conceito.

###Conclusão

&emsp;Conforme o dataframe acima, é possível perceber que aumentando o valor da janela do PSI para além do que foi inicialmente estipulado como padrão, janela de tamanho 100, o método é capaz de detectar com maior precisão a quantidade de mudanças ocorridas. \
&emsp;Podemos notar que o PSI obteve melhores resultados que os seguintes detectores no quesito de acurácia: EDDM, ADWIN e PAGE-HINKLEY, esse inclusive deixou de detectar mudanças que ocorreram na stream. Quanto ao atraso para a detecção, ficou somente atrás do ADWIN e CUSUM, respectivamente.\
&emsp;Como próximos passos, é preciso validar os testes realizados para replicar os resultados, podendo assim comprovar a eficácia do PSI e seguir para a próxima sprint.